include/nbl/asset/filters/CFlattenRegionsStreamHashImageFilter.h

```diff
@@ -0,0 +1,220 @@
+
+					assert(memsize.getNumerator() % memsize.getDenominator() == 0u);
+					bufferSize += memsize.getIntegerApprox();
+				}
+			}
+
+			auto executePerMipLevel = [&](const uint32_t miplevel)
+			{
+				/*
+					we stream-hash texels per given mip level & layer
+				*/
+
+				auto* hashers = _NBL_NEW_ARRAY(blake3_hasher, parameters.arrayLayers);
+
+				for (auto layer = 0u; layer < parameters.arrayLayers; ++layer)
+					blake3_hasher_init(&hashers[layer]);
+
+				auto hash = [&hashers, &inData, &texelOrBlockByteSize](uint32_t readBlockArrayOffset, core::vectorSIMDu32 readBlockPos) -> void
+				{
+					auto* hasher = hashers + readBlockPos.w;
+					blake3_hasher_update(hasher, inData + readBlockArrayOffset, texelOrBlockByteSize);
+				};
+
+				IImage::SSubresourceLayers subresource = { .aspectMask = static_cast<IImage::E_ASPECT_FLAGS>(0u), .mipLevel = miplevel, .baseArrayLayer = 0u, .layerCount = parameters.arrayLayers }; // stick to given mip level and take all layers
+				CMatchedSizeInOutImageFilterCommon::state_type::TexelRange range = { .offset = {}, .extent = { parameters.extent.width, parameters.extent.height, parameters.extent.depth } }; // cover all texels within layer range, take 0th mip level size to not clip anything at all
+				CBasicImageFilterCommon::clip_region_functor_t clipFunctor(subresource, range, parameters.format);
+
+				CBasicImageFilterCommon::executePerRegion(policy, proxy.flatten.outImage.get(), hash, regions->begin(), regions->end(), clipFunctor); // fire the hasher for layers with specified execution policy, yes you can use parallel policy here if you want at it will work
+
+				for (auto layer = 0u; layer < parameters.arrayLayers; ++layer)
+				{
+					auto* out = reinterpret_cast<uint8_t*>(reinterpret_cast<CState::hash_t*>(state->scratchMemory.heap->getPointer()) + (miplevel * parameters.arrayLayers) + layer);
+					blake3_hasher_finalize(hashers + layer, out, sizeof(CState::hash_t)); // finalize hash for layer + put it to heap for given mip level
+				}
+
+				_NBL_DELETE_ARRAY(hashers, parameters.arrayLayers);
+			};
+
+			std::vector<uint32_t> levels(parameters.mipLevels);
+			std::iota(levels.begin(), levels.end(), 0);
+
+			std::for_each(policy, levels.begin(), levels.end(), executePerMipLevel); // fire per block of layers for given mip level with specified execution policy, yes you can use parallel policy here if you want at it will work
+
+			/*
+				scratch's heap is filled with all hashes, 
+				time to use them and compute final hash
+			*/
+
+			blake3_hasher hasher;
+			blake3_hasher_init(&hasher);
+			{
+				for (auto miplevel = 0u; miplevel < parameters.mipLevels; ++miplevel)
+					for (auto layer = 0u; layer < parameters.arrayLayers; ++layer)
+					{
+						auto* hash = reinterpret_cast<CState::hash_t*>(state->scratchMemory.heap->getPointer()) + (miplevel * parameters.arrayLayers) + layer;
+						blake3_hasher_update(&hasher, hash->data(), sizeof(CState::hash_t));
+					}
```

Comment:
here take hashes of mip levels and turn into one hash

Reply:
does it make any difference? It will make my scratch smaller but I would like to store each hash in the scratch for our [tool & testing purposes](https://github.com/Devsh-Graphics-Programming/Nabla-Examples-and-Tests/tree/14b34ffab28112de092b56d17be52e32d450b834/64_StreamHash#streamhashapp) (however I could add template parameter which toggles whether I store all hashes or only for mips to do what you have in mind - by default would do it)

Reply:
I leave it how it is, want to keep all data in scratch unless you want template I can add it